homeassistant/components/bmw_connected_drive/lock.py

```diff
@@ -60,38 +57,43 @@ def __init__(
 
     def lock(self, **kwargs: Any) -> None:
         """Lock the car."""
-        _LOGGER.debug("%s: locking doors", self._vehicle.name)
-        # Optimistic state set here because it takes some time before the
-        # update callback response
-        self._attr_is_locked = True
-        self.schedule_update_ha_state()
-        self._vehicle.remote_services.trigger_remote_door_lock()
+        _LOGGER.debug("%s: locking doors", self.vehicle.name)
+        # Only update the HA state machine if the vehicle reliably reports its lock state
+        if self.door_lock_state_available:
+            # Optimistic state set here because it takes some time before the
+            # update callback response
+            self._attr_is_locked = True
+            self.schedule_update_ha_state()
+        self.vehicle.remote_services.trigger_remote_door_lock()
 
     def unlock(self, **kwargs: Any) -> None:
         """Unlock the car."""
-        _LOGGER.debug("%s: unlocking doors", self._vehicle.name)
-        # Optimistic state set here because it takes some time before the
-        # update callback response
-        self._attr_is_locked = False
-        self.schedule_update_ha_state()
-        self._vehicle.remote_services.trigger_remote_door_unlock()
-
-    def update(self) -> None:
-        """Update state of the lock."""
-        _LOGGER.debug(
-            "Updating lock data for '%s' of %s", self._attribute, self._vehicle.name
-        )
-        vehicle_state = self._vehicle.status
-        if not self.door_lock_state_available:
-            self._attr_is_locked = None
-        else:
+        _LOGGER.debug("%s: unlocking doors", self.vehicle.name)
+        # Only update the HA state machine if the vehicle reliably reports its lock state
+        if self.door_lock_state_available:
+            # Optimistic state set here because it takes some time before the
+            # update callback response
+            self._attr_is_locked = False
+            self.schedule_update_ha_state()
+        self.vehicle.remote_services.trigger_remote_door_unlock()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        _LOGGER.debug("Updating lock data of %s", self.vehicle.name)
+        # Only update the HA state machine if the vehicle reliably reports its lock state
+        if self.door_lock_state_available:
+            vehicle_state = self.vehicle.status
             self._attr_is_locked = vehicle_state.door_lock_state in {
                 LockState.LOCKED,
                 LockState.SECURED,
             }
+            self._attr_extra_state_attributes = dict(
+                self._attrs,
+                **{
+                    "door_lock_state": vehicle_state.door_lock_state.value,
+                    "last_update_reason": vehicle_state.last_update_reason,
```

Comment:
It looks like these should be `sensor` entities
https://developers.home-assistant.io/docs/core/entity/?_highlight=extra_state_attributes

Reply:
I know, sorry. It regularly comes up but haven't had the chance to go give them a good clean up.

Reply:
Thanks for the reminder, thats on my todo list once the bigger backend changes are done!